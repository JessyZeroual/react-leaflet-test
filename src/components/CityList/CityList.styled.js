import styled from "styled-components";
import { CityItemHeight } from "../../constant/style-constants";

export const StyledCityItem = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: ${CityItemHeight}px;
  margin: 10px;
  background: #fff;
  box-shadow: ${({ active }) => active && "0px 2px 4px grey"};
  border-radius: 5px;
  transform-origin: bottom;
  transition: all 0.1s ease-in-out;


  &:hover {
    cursor: pointer;
    box-shadow: 0px 1px 3px grey;
    transform: scale(1.05);
  }
`;

