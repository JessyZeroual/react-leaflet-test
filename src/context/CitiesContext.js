import React from "react";

const CitiesContext = React.createContext({});

export const CitiesContextProvider = CitiesContext.Provider;
export const CitiesContextConsumer = CitiesContext.Consumer;
export default CitiesContext;
