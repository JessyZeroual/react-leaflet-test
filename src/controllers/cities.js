export const getCities = () => {
  return fetch(`/cities.json`)
    .then((res) => res.json())
    .then((data) =>
      data.map((city) => ({
        ...city,
        growth_from_2000_to_2013: Number(
          city.growth_from_2000_to_2013.replace("%", "")
        ),
        population: Number(city.population),
        rank: Number(city.rank),
      }))
    );
};
