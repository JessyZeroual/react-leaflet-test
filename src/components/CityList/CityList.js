import React, { useContext } from "react";
import ClipLoader from "react-spinners/ClipLoader";

import CitiesContext from "../../context/CitiesContext";
import useInfiniteScroll from "../../utils/useInfiniteScroll";
import { secondaryColor } from "../../constant/style-constants";
import { StyledCityItem } from "./CityList.styled";
import { StyledWrapperCenter } from "../../style/common.styled-component";

const CityList = () => {
  const citiesContext = useContext(CitiesContext);
  const { isLoading, cities, currentCity } = citiesContext.state;
  const { setCurrentCity } = citiesContext.dispatch;

  const {
    numberOfElements,
    fetchingMoreElements,
    hasNextElements,
  } = useInfiniteScroll(cities);

  return (
    <>
      {isLoading ? (
        <StyledWrapperCenter>
          <ClipLoader size={40} color={secondaryColor} />
        </StyledWrapperCenter>
      ) : (
        <>
          {cities.slice(0, numberOfElements).map((city, key) => (
            <StyledCityItem
              key={key}
              active={currentCity.city === city.city}
              onClick={() => setCurrentCity(city)}
            >
              {city.city}
            </StyledCityItem>
          ))}
          {fetchingMoreElements && (
            <StyledWrapperCenter>
              <ClipLoader size={20} color={secondaryColor} />
            </StyledWrapperCenter>
          )}
          {!hasNextElements && (
            <StyledWrapperCenter>No other cities available</StyledWrapperCenter>
          )}
        </>
      )}
    </>
  );
};

export default CityList;
