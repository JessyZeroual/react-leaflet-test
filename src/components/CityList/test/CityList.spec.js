import React from "react";
import TestRenderer from "react-test-renderer";
import { act } from "react-test-renderer";

import CityList from "../CityList";
import CitiesContext from "../../../context/CitiesContext";
import cities from "../../../test/listOfCitiesTests";

const elementMock = {
  addEventListener: jest.fn(),
  removeEventListener: jest.fn(),
};
jest.spyOn(document, "getElementById").mockImplementation(() => elementMock);

const currentCity = [
  {
    city: "New York",
    growth_from_2000_to_2013: "4.8%",
    latitude: 40.7127837,
    longitude: -74.0059413,
    population: "8405837",
    rank: "1",
    state: "New York",
  },
];

const setCurrentCity = jest.fn();

describe("CityList", () => {
  let renderer;

  const mountCityList = () =>
    act(() => {
      renderer = TestRenderer.create(
        <CitiesContext.Provider
          value={{
            state: { isLoading: false, cities, currentCity },
            dispatch: { setCurrentCity },
          }}
        >
          <CityList />
        </CitiesContext.Provider>
      );
    });

  describe("When CityList is mounted", () => {
    it("renders correctly", () => {
      mountCityList();
      expect(renderer).toMatchSnapshot();
    });

    it("receives a list of 20 cities", () => {
      mountCityList();
      expect(renderer.toJSON().length).toEqual(20);
    });

    describe("When we scroll down", () => {
      mountCityList();
      it("adds 20 cities to the list", async () => {
        await act(async () => {
          // Simulate scroll to bottom
        });
        // expect(renderer.toJSON().length).toEqual(40);
      });
    });
  });
});
