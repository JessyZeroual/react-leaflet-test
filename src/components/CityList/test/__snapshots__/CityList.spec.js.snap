// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityList When CityList is mounted renders correctly 1`] = `
Array [
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    New York
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    Los Angeles
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    Chicago
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    Houston
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    Philadelphia
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    Phoenix
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    San Antonio
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    San Diego
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    Dallas
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    San Jose
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    Austin
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    Indianapolis
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    Jacksonville
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    San Francisco
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    Columbus
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    Charlotte
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    Fort Worth
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    Detroit
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    El Paso
  </div>,
  <div
    className="sc-AxjAm bbjfxF"
    onClick={[Function]}
  >
    Memphis
  </div>,
]
`;
